sdrf_pipelines/sdrf/sdrf.py

```diff
@@ -77,3 +79,121 @@ def validate(self, template: str):
+          raised.
+        :return: A list of LogicError objects if the experimental design is incorrect, otherwise an empty list.
+        """
+
+        errors = []
+
+        # Check that combination of value assay name and characteristics[data file] is unique in self
+        errors = self.check_inconsistencies_assay_file(errors)
+
+        errors = self.check_unique_sample_file_combinations(errors)
+
+        return errors
+
+    def check_inconsistencies_assay_file(self, errors: List[LogicError]) -> List[LogicError]:
+        """
+        Check that combination of values assay name and comment[data file] is unique in self.
+        :return: A list of LogicError objects if the combination of values assay name and characteristics[data file] is
+        not unique, otherwise an empty list.
+        """
+
+        # Group by col1 and check if each group has only one unique col2 value
+        col1_inconsistencies = self.groupby("assay name")["comment[data file]"].nunique()
+        col1_inconsistent_groups = col1_inconsistencies[col1_inconsistencies > 1]
+        if len(col1_inconsistent_groups) > 0:
+            cell_index = col1_inconsistent_groups.index.tolist()
+            error_message = f"Multiple assays with the same raw files: {cell_index}, the combination assay name and comment[data file] should be unique"
+            errors.append(LogicError(error_message, error_type=logging.ERROR))
+
+        # Group by col2 and check if each group has only one unique col1 value
+        col2_inconsistencies = self.groupby("comment[data file]")["assay name"].nunique()
+        col2_inconsistent_groups = col2_inconsistencies[col2_inconsistencies > 1]
+        if len(col2_inconsistent_groups) > 0:
+            cell_index = col2_inconsistent_groups.index.tolist()
+            error_message = f"Multiple raw files with the same assay: {cell_index}, the combination assay name and comment[data file] should be unique"
+            errors.append(LogicError(error_message, error_type=logging.ERROR))
+
+        return errors
+
+    def check_unique_sample_file_combinations(self, errors: List[LogicError]) -> List[LogicError]:
+        """
+        The combination of the following columns should be unique:
+        - source name
+        - comment[technical replicate]
+        - comment[biological replicate]
+        - comment[label]
+        - comment[fraction identifier]
+        :return: A list of LogicError objects if the source names are not unique, otherwise an empty list.
+        """
+        cols = [
+            "source name",
+            "comment[technical replicate]",
+            "characteristics[biological replicate]",
+            "comment[label]",
+            "comment[fraction identifier]",
+        ]
+
+        for col in cols:
+            if col not in self.columns:
+                error_message = (
+                    f"In order to perform experimental design validation, column '{col}' must be present in the SDRF"
```

Comment:
I think that's a bit too restrictive. Wouldn't one assume that if e.g. no technical or biological replicate is given, that all of them are different ones?
Can we not imply this? Sdrfs are already verbose enough 😅

Reply:
For an SDRF to be valid the technical and biological replicates MUST be provided.